plutus-core/plutus-core/src/PlutusCore/Builtin/KnownTypeAst.hs

```diff
@@ -143,77 +150,101 @@ data family RepHole x
 data family TypeHole a
 
--- | For annotating an uninstantiated built-in type, so that it gets handled by the right instance
--- or type family.
-type BuiltinHead :: forall k. k -> k
-data family BuiltinHead f
-
--- | Take an iterated application of a built-in type and elaborate every function application
--- inside of it to 'TypeAppRep', plus annotate the head with 'BuiltinHead'.
--- The idea is that we don't need to process built-in types manually if we simply add some
--- annotations for instance resolution to look for. Think what we'd have to do manually for, say,
--- 'ToHoles': traverse the spine of the application and collect all the holes into a list, which is
--- troubling, because type applications are left-nested and lists are right-nested, so we'd have to
--- use accumulators or an explicit 'Reverse' type family. And then we also have 'KnownTypeAst' and
--- 'ToBinds', so handling built-in types in a special way for each of those would be a hassle,
--- especially given the fact that type-level Haskell is not exactly good at computing things.
--- With the 'ElaborateBuiltin' approach we get 'KnownTypeAst', 'ToHoles' and 'ToBinds' for free.
-type ElaborateBuiltin :: forall k. k -> k
-type family ElaborateBuiltin a where
-    ElaborateBuiltin (f x) = ElaborateBuiltin f `TyAppRep` x
-    ElaborateBuiltin f     = BuiltinHead f
+{- Note [Name generality of KnownTypeAst]
+The 'KnownTypeAst' class takes a @tyname@ argument. The reason for this is that we want to be able
+to define 'mkTyBuiltin' such that it's generic over the type of names, because in addition to
+'TyName' we also have 'TyDeBruijn' and it's convenient to seemlessly embed built-in types into a
+'Type' regardless of which kind of names it expects.
+
+However we don't want that @tyname@ to proliferate through the entire code base. For example,
+'HasTypeLevel' propagates up to PlutusTx and it doesn't make sense to mention implementation details
+such as 'TyName' there. For this reason we instantiate @tyname@ as soon as possible. When we want
+and can have full generality, we instantiate @tyname@ with 'Void', because anything can be recovered
+from 'Void' via 'absurd' while allowing us not to thread the @tyname@ parameter through half the
+codebase. And when we don't want or can't have generality, we instantiate @tyname@ with 'TyName'.
+-}
 
--- | A constraint for \"@a@ is a 'KnownTypeAst' by means of being included in @uni@\".
-type KnownBuiltinTypeAst uni a = KnownTypeAst uni (ElaborateBuiltin a)
+-- | Specifies that the given type is a built-in one and can be embedded into a 'Type'.
+type HasTypeLevel :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+type HasTypeLevel uni x = KnownTypeAst Void uni (ElaborateBuiltin uni x)
+
+-- | Specifies that the given type is a built-in one and its values can be embedded into a 'Term'.
+type HasTermLevel :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+type HasTermLevel uni = Includes uni
```

Comment:
> Could we put this in there instead and not export `Includes` at all? 

Well, `Universe.Core` doesn't know anything about the type-level vs term-level distinction, which is why I didn't want to introduce `HasTermLevel` there. 

> Also, if `Includes` is essentially confined to `Universe.Core` it might be possible to give it a longer but more illuminating name.

Agreed, I've created PLT-6851 for this.

Keep in mind that I'm going to change `Universe.Core` a lot too. The whole `Esc`, `Kinded` etc business is going to be removed simplifying the library. Unfortunately, `Includes` is there to stay, but maybe we'll decide to introduce the type-level vs term-level distinction right into the library (it's our internal library anyway) and replace `Includes` with `HasTermLevel`, although I'm not fond of this, because I think `Includes` is more general than that.